Source/WebCore/platform/graphics/BitmapImage.cpp

```diff
@@ -27,618 +27,143 @@
-        orientation = frameOrientationAtIndex(m_currentFrame);
-        drawNativeImage(*image, context, destRect, srcRect, IntSize(sourceSize(orientation)), { options, orientation });
-    } else
-        drawNativeImage(*image, context, destRect, srcRect, IntSize(sourceSize(orientation)), options);
+    if (auto color = (*nativeImage)->singlePixelSolidColor())
+        fillWithSolidColor(context, destinationRect, *color, options.compositeOperator());
+    else {
+        // adjustedSourceRect is in the coordinates of the unsubsampled image, so map it to the subsampled image.
+        auto imageSize = (*nativeImage)->size();
+        if (imageSize != sourceSize)
+            adjustedSourceRect.scale(imageSize / sourceSize);
+
+        auto orientation = options.orientation();
+        if (orientation == ImageOrientation::Orientation::FromImage)
+            orientation = currentFrameOrientation();
 
-    m_currentFrameDecodingStatus = frameDecodingStatusAtIndex(m_currentFrame);
+        context.drawNativeImage(*nativeImage, destinationRect, adjustedSourceRect, { options, orientation });
+    }
 
     if (auto observer = imageObserver())
         observer->didDraw(*this);
 
-    return result;
+    return ImageDrawResult::DidDraw;
 }
 
-void BitmapImage::drawPattern(GraphicsContext& ctxt, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& transform, const FloatPoint& phase, const FloatSize& spacing, ImagePaintingOptions options)
+void BitmapImage::drawPattern(GraphicsContext& context, const FloatRect& destinationRect, const FloatRect& tileRect, const AffineTransform& transform, const FloatPoint& phase, const FloatSize& spacing, ImagePaintingOptions options)
 {
     if (tileRect.isEmpty())
         return;
 
-    if (!ctxt.drawLuminanceMask()) {
-        // If new data is received, the current incomplete decoded frame has to be destroyed.
-        if (m_currentFrameDecodingStatus == DecodingStatus::Invalid)
-            m_source->destroyIncompleteDecodedData();
-
-        Image::drawPattern(ctxt, destRect, tileRect, transform, phase, spacing, { options, ImageOrientation::Orientation::FromImage });
-        m_currentFrameDecodingStatus = frameDecodingStatusAtIndex(m_currentFrame);
-        return;
-    }
-
-    if (!m_cachedImage) {
-        auto buffer = ctxt.createAlignedImageBuffer(expandedIntSize(tileRect.size()));
-        if (!buffer)
-            return;
-
-        auto observer = imageObserver();
-
-        // Temporarily reset image observer, we don't want to receive any changeInRect() calls due to this relayout.
-        setImageObserver(nullptr);
-
-        draw(buffer->context(), tileRect, tileRect, { options, DecodingMode::Synchronous, ImageOrientation::Orientation::FromImage });
-
-        setImageObserver(WTFMove(observer));
-        buffer->convertToLuminanceMask();
-
-        m_cachedImage = ImageBuffer::sinkIntoNativeImage(WTFMove(buffer));
-        if (!m_cachedImage)
-            return;
```

Comment:
We no longer store this cached NativeImage when drawing patterns. Is it unnecessary?

Reply:
This code does not look correct. We create the `ImageBuffer` with size = tileRect.size(). And we cache this `ImageBuffer` to be used for any subsequent drawing with for `drawLuminanceMask()` equals true. But we do not check whether the current `tileRect.size()` equals the size of the ImageBuffer or not. Besides it seems `drawLuminanceMask()` equals true a rare case.